File: firmware/src/sht3x.cpp
#include "sht3x.h"

#include <Arduino.h>
#include <Wire.h>

#include "config.h"

namespace {

// CRC-8, polynomial 0x31, init 0xFF — identical to the Python _crc8().
uint8_t crc8(const uint8_t* data, int len) {
  uint8_t crc = 0xFF;
  for (int i = 0; i < len; ++i) {
    crc ^= data[i];
    for (int b = 0; b < 8; ++b) {
      crc = (crc & 0x80) ? static_cast<uint8_t>((crc << 1) ^ 0x31)
                         : static_cast<uint8_t>(crc << 1);
    }
  }
  return crc;
}

}  // namespace

void SHT3x::begin() {
  Wire.begin(PIN_I2C_SDA, PIN_I2C_SCL, I2C_HZ);
}

bool SHT3x::read(Reading& out) {
  for (uint8_t attempt = 0; attempt < retries_; ++attempt) {
    if (readOnce(out)) return true;
    delay(50 * (attempt + 1));  // incremental backoff: 50, 100, 150 ms
  }
  return false;
}

bool SHT3x::readOnce(Reading& out) {
  // Trigger a single high-repeatability measurement: command 0x2C 0x06.
  Wire.beginTransmission(addr_);
  Wire.write(0x2C);
  Wire.write(0x06);
  if (Wire.endTransmission() != 0) return false;

  delay(20);  // high-rep conversion completes well under 20 ms

  if (Wire.requestFrom(static_cast<int>(addr_), 6) != 6) return false;
  uint8_t d[6];
  for (int i = 0; i < 6; ++i) d[i] = Wire.read();

  // Two CRC-protected words: [temp_hi, temp_lo, crc][hum_hi, hum_lo, crc].
  if (crc8(d, 2) != d[2] || crc8(d + 3, 2) != d[5]) return false;

  const uint16_t t_raw = (static_cast<uint16_t>(d[0]) << 8) | d[1];
  const uint16_t h_raw = (static_cast<uint16_t>(d[3]) << 8) | d[4];
  out.temp_c = -45.0f + 175.0f * t_raw / 65535.0f;
  out.humidity = 100.0f * h_raw / 65535.0f;
  return true;
}

File: milkv_duo/sht3x.py
"""SHT3x reader for the Milk-V Duo using only the stdlib I2C char device.

The Pi build uses smbus2, which is not on the stock Duo image. Here we talk to
/dev/i2c-N directly with an I2C_SLAVE ioctl plus os.read/os.write -- enough for
the SHT3x single-shot measurement (write command, wait, read 6 bytes). The
CRC-8 verification matches app/sensor.py on the Pi.
"""

from __future__ import annotations

import fcntl
import os
import time
from dataclasses import dataclass

I2C_SLAVE = 0x0703
SENSOR_ADDR = 0x44
CMD_SINGLE_HIGHREP = bytes((0x2C, 0x06))
MEASUREMENT_DELAY_S = 0.020


class SensorError(Exception):
    pass


@dataclass(frozen=True)
class Reading:
    temp_c: float
    humidity: float
    ts: float


def _crc8(data: bytes) -> int:
    crc = 0xFF
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = ((crc << 1) ^ 0x31) & 0xFF if crc & 0x80 else (crc << 1) & 0xFF
    return crc


class Sensor:
    """Reads an SHT3x on /dev/i2c-{bus_num}. Retries like the Pi driver."""

    def __init__(self, bus_num: int = 0, address: int = SENSOR_ADDR,
                 retries: int = 3) -> None:
        self._dev = f"/dev/i2c-{bus_num}"
        self._addr = address
        self._retries = retries

    def read(self) -> Reading:
        last_err: Exception | None = None
        for attempt in range(self._retries):
            try:
                return self._read_once()
            except (OSError, SensorError) as e:
                last_err = e
                time.sleep(0.05 * (attempt + 1))
        raise SensorError(f"sensor read failed after {self._retries} attempts: {last_err}")

    def _read_once(self) -> Reading:
        fd = os.open(self._dev, os.O_RDWR)
        try:
            fcntl.ioctl(fd, I2C_SLAVE, self._addr)
            os.write(fd, CMD_SINGLE_HIGHREP)
            time.sleep(MEASUREMENT_DELAY_S)
            data = os.read(fd, 6)
        finally:
            os.close(fd)

        if len(data) != 6:
            raise SensorError("short read")
        if _crc8(data[0:2]) != data[2] or _crc8(data[3:5]) != data[5]:
            raise SensorError("CRC mismatch")

        t_raw = (data[0] << 8) | data[1]
        h_raw = (data[3] << 8) | data[4]
        return Reading(
            temp_c=-45.0 + 175.0 * t_raw / 65535.0,
            humidity=100.0 * h_raw / 65535.0,
            ts=time.time(),
        )


class FakeSensor:
    """Deterministic-ish reading for --sim runs with no I2C bus."""

    def __init__(self) -> None:
        self._t0 = time.time()

    def read(self) -> Reading:
        import math

        dt = time.time() - self._t0
        return Reading(
            temp_c=4.0 + 1.5 * math.sin(dt / 6.0),
            humidity=60.0 + 5.0 * math.sin(dt / 11.0),
            ts=time.time(),
        )
